vpr/src/pack/appack_max_dist_th_manager.cpp

```diff
@@ -0,0 +1,276 @@
+/**
+ * @file
+ * @author  Alex Singer
+ * @date    May 2025
+ * @breif   Definition of the max distance threshold manager class.
+ */
+
+#include "appack_max_dist_th_manager.h"
+#include <regex>
+#include <stdexcept>
+#include <tuple>
+#include <vector>
+#include "device_grid.h"
+#include "physical_types.h"
+#include "physical_types_util.h"
+#include "vpr_error.h"
+#include "vpr_utils.h"
+#include "vtr_assert.h"
+#include "vtr_log.h"
+
+/**
+ * @brief Helper method to convert a string into a float with error checking.
+ */
+static float str_to_float_or_error(const std::string& str);
+
+/**
+ * @brief Helper method to parse one term of the user-provided max distance
+ *        threshold string.
+ *
+ * This method decomposes the user string of the form "<regex>:<scale>,<offset>"
+ * into its three components.
+ */
+static std::tuple<std::string, float, float> parse_max_dist_th(const std::string& max_dist_th);
+
+/**
+ * @brief Recursive helper method to deduce if the given pb_type is or contains
+ *        pb_types which are of the memory class.
+ *
+ * TODO: This should be a graph traversal instead of a recursive function.
+ */
+static bool has_memory_pbs(const t_pb_type* pb_type);
+
+APPackMaxDistThManager::APPackMaxDistThManager(bool should_initialize,
+                                               const std::vector<std::string>& max_dist_ths,
+                                               const std::vector<t_logical_block_type>& logical_block_types,
+                                               const DeviceGrid& device_grid) {
+    // Check if we should initialize or not. If not, just exit.
+    if (!should_initialize) {
```

Comment:
I'm not sure if it's common practice to pass a flag to a constructor to determine whether a class should be initialized or not. I think it's a bit unusual. There are other classes in VPR that follow a different pattern, where an initialize() method is called explicitly after construction. I tend to prefer that approach, but I’ll leave it to your judgment.